tsdb/index/labelvalues.go

```diff
@@ -0,0 +1,332 @@
+}
+
+func (it *intersectLabelValuesV1) Err() error {
+	return it.err
+}
+
+func (it *intersectLabelValuesV1) Warnings() annotations.Annotations {
+	return nil
+}
+
+func (it *intersectLabelValuesV1) Close() error {
+	return nil
+}
+
+type intersectLabelValues struct {
+	d         *encoding.Decbuf
+	b         ByteSlice
+	dec       *Decoder
+	postings  *PostingsCloner
+	lastVal   string
+	skip      int
+	cur       string
+	exhausted bool
+	err       error
+	inverted  bool
+}
+
+func (it *intersectLabelValues) Next() bool {
+	if it.exhausted || it.err != nil {
+		return false
+	}
+
+	for !it.exhausted && it.d.Err() == nil {
+		if it.skip == 0 {
+			// These are always the same number of bytes,
+			// and it's faster to skip than to parse.
+			it.skip = it.d.Len()
+			// Key count
+			it.d.Uvarint()
+			// Label name
+			it.d.UvarintBytes()
+			it.skip -= it.d.Len()
+		} else {
+			it.d.Skip(it.skip)
+		}
+
+		// Label value
+		vb := it.d.UvarintBytes()
+		v := yoloString(vb)
+
+		postingsOff := int(it.d.Uvarint64())
+		// Read from the postings table
+		postingsDec := encoding.NewDecbufAt(it.b, postingsOff, castagnoliTable)
+		_, curPostings, err := it.dec.Postings(postingsDec.Get())
+		if err != nil {
+			it.err = fmt.Errorf("decode postings: %w", err)
+			return false
+		}
+
+		it.exhausted = v == it.lastVal
+
+		isMatch := intersect(curPostings, it.postings.Clone())
+		if it.inverted {
+			isMatch = !isMatch
+		}
+		if isMatch {
+			// Make sure to allocate a new string
+			it.cur = string(vb)
+			return true
+		}
+	}
+	if it.d.Err() != nil {
+		it.err = fmt.Errorf("get postings offset entry: %w", it.d.Err())
+	}
+
+	return false
+}
+
+func (it *intersectLabelValues) At() string {
+	return it.cur
+}
+
+func (it *intersectLabelValues) Err() error {
+	return it.err
+}
+
+func (it *intersectLabelValues) Warnings() annotations.Annotations {
+	return nil
+}
+
+func (it *intersectLabelValues) Close() error {
+	return nil
+}
+
+// LabelValuesFor returns LabelValues for the given label name in the series referred to by postings.
+func (p *MemPostings) LabelValuesFor(postings Postings, name string) storage.LabelValues {
+	return p.labelValuesFor(postings, name, false)
+}
+
+// LabelValuesNotFor returns LabelValues for the given label name in the series *not* referred to by postings.
+func (p *MemPostings) LabelValuesNotFor(postings Postings, name string) storage.LabelValues {
+	return p.labelValuesFor(postings, name, true)
+}
+
+func (p *MemPostings) labelValuesFor(postings Postings, name string, inverted bool) storage.LabelValues {
+	p.mtx.RLock()
+
+	e := p.m[name]
+	if len(e) == 0 {
+		p.mtx.RUnlock()
+		return storage.EmptyLabelValues()
+	}
+
+	// With thread safety in mind and due to random key ordering in map, we have to construct the array in memory
+	vals := make([]string, 0, len(e))
+	candidates := make([]Postings, 0, len(e))
+	for val, srs := range e {
+		vals = append(vals, val)
+		candidates = append(candidates, NewListPostings(srs))
+	}
```

Comment:
There's one allocation happening for each label value, and we can easily avoid that:

```suggestion
	candidates := make([]Postings, 0, len(e))
	// Allocate a slice for all needed ListPostings, so no need to allocate them one by one.
	lps := make([]ListPostings, 0, len(e))
	for val, srs := range e {
		vals = append(vals, val)
		lps = append(lps, ListPostings{list: srs})
		candidates = append(candidates, &lps[len(lps)-1])
	}
```

Since `ListPostings` is part of this package, it should be fine to instantiate it without using the constructor.